File: 0130-surrounded-regions/0130-surrounded-regions.java
class Solution {
    public void solve(char[][] board) {
        int r=board.length;
        int c = board[0].length;
        for(int i=0;i<r;i++)
        {
            for(int j=0;j<c;j++)
            {
                if(board[i][j] == 'O' && (i == 0 || i == r - 1 || j == 0 || j == c - 1))
                {
                        change(board,i,j);
                }
            }
        }
        for(int i=0;i<r;i++)
        {
            for(int j=0;j<c;j++)
            {
                if(board[i][j] == 'O')
                    board[i][j] = 'X';
                if(board[i][j]=='R')
                    board[i][j] = 'O';
            }
        }
    }
    void change(char[][] board,int r,int c)
    {
        if(r<0 || c<0 || r>=board.length || c>=board[0].length || board[r][c]!='O')
            return;
        board[r][c] = 'R';
        change(board,r-1,c);
        change(board,r+1,c);
        change(board,r,c-1);
        change(board,r,c+1);
    }
}

File: 0130-surrounded-regions/0130-surrounded-regions.py
class Solution:
    def solve(self, board: list[list[str]]) -> None:
        if not board or not board[0]:
            return

        r = len(board)
        c = len(board[0])

        # Step 1: Mark boundary 'O's and their connected 'O's as 'R'
        for i in range(r):
            for j in range(c):
                if board[i][j] == 'O' and (i == 0 or i == r - 1 or j == 0 or j == c - 1):
                    self.change(board, i, j)

        # Step 2: Flip unvisited 'O' to 'X' and revert 'R' back to 'O'
        for i in range(r):
            for j in range(c):
                if board[i][j] == 'O':
                    board[i][j] = 'X'
                elif board[i][j] == 'R':
                    board[i][j] = 'O'

    def change(self, board: list[list[str]], r: int, c: int) -> None:
        if (r < 0 or c < 0 or 
            r >= len(board) or c >= len(board[0]) or 
            board[r][c] != 'O'):
            return

        board[r][c] = 'R'
        self.change(board, r - 1, c)
        self.change(board, r + 1, c)
        self.change(board, r, c - 1)
        self.change(board, r, c + 1)
